opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -87,23 +87,50 @@ message InstrumentationLibraryMetrics {
 //                       +---------------------------+
 //
 //-----------------------------------------------------------------------
-// DataPoint is a value of specific type corresponding to a given moment in
-// time. Each DataPoint is timestamped.
-//
-// DataPoint is strongly typed: each DataPoint type has a specific Protobuf message
-// depending on the value type of the metric and thus there are currently 4 DataPoint
-// messages, which correspond to the types of metric values.
 message Metric {
   // metric_descriptor describes the Metric.
   MetricDescriptor metric_descriptor = 1;
 
-  // Data is a list of one or more DataPoints for a single metric. Only one of the
-  // following fields is used for the data, depending on the type of the metric defined
-  // by MetricDescriptor.type field.
-  repeated Int64DataPoint int64_data_points = 2;
-  repeated DoubleDataPoint double_data_points = 3;
-  repeated HistogramDataPoint histogram_data_points = 4;
-  repeated SummaryDataPoint summary_data_points = 5;
+  repeated Data data = 2;
+
+  message Data {
+    // The set of labels that uniquely identify this timeseries.
+    //
+    // These labels take precedence over any labels with the same key defined
+    // in the metric descriptor.
+    repeated opentelemetry.proto.common.v1.StringKeyValue labels = 1;
+
+    // start_time_unix_nano is the beginning of a time interval over which
+    // this data point was recorded. It MUST be represented as the UNIX Epoch
+    // time in nanoseconds (nanoseconds since 00:00:00 UTC on 1 January 1970).
+    // If zero, it is treated as unspecified and may be set by the receiver.
+    //
+    // If this data point is a part of a timeseries for an INSTANTANEOUS
+    // metric, this value is ignored.
+    //
+    // If this data point is a part of a timeseries for an DELTA, or
+    // CUMULATIVE metric, this value represents the instant before the
+    // measuring time interval began (i.e. for time t to be in the measuring
+    // interval than start_time_unix_nano < t <= time_unix_nano).
```

Comment:
This `start < t <= time` bothers me.  Why not `start <= t < time`? I state this because it's possible for an event to happen concurrently with the start of a time window and be recorded in that window, but it's not possible for an event to happen concurrently with the end of a window and be included in the window.

Reply:
the way how "time" is implemented in the current systems cannot guarantee a happens-before relationship so I would not bother to justify what should happen if an event happens at that same moment - unless there is another way to synchronize these events time cannot guarantee the ordering of the events.

> I state this because it's possible for an event to happen concurrently with the start of a time window and be recorded in that window, but it's not possible for an event to happen concurrently with the end of a window and be included in the window.

If an event happens at the same time (if you read `time.Now` exactly when the event happens) with closing the time we close the window (if you read `time.Now`) there is no guarantee if the event will be included in the current window or the next window unless there are other mechanisms to ensure the happens before relationship, so none of the option is ideal.

The idea behind using this definition of the (start, end] is that when using an Observer the value is "added" at the end time so it feels kind of more natural this way - again none of the versions can be 100% correct, so we also look at what others do like OpenMetrics/Stackdriver those being the system which have start time.

Reply:
> we also look at what others do like OpenMetrics/Stackdriver

@bogdandrutu : where is this defined for OpenMetrics?

Reply:
FWIW, Stackdriver is an inclusive interval: https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeInterval

E.g. `[start, end]`

Reply:
Looking at a token example of a check-pointing process:

```
=========================
||     Checkpointer    ||
=========================


-------------------------
| Start: t_0            |
-------------------------
       |                            -----------------
       |  <-------------------------| measurement_0 |
       |                            -----------------
-------------------------
| Checkpoint 1          |
| --------------------- |           -----------------
| Begin Checkpoint: t_1 |  <--------| measurement_1 |
|                       |           -----------------
| Processing...         |
|                       |           -----------------
|                       |  <--------| measurement_2 |
|                       |           -----------------
-------------------------
       |                            -----------------
       |  <-------------------------| measurement_3 |
       |                            -----------------
-------------------------
| Checkpoint 2          |
| --------------------- |
| Begin Checkpoint: t_2 |
|                       |
| Processing...         |
|                       |
-------------------------
       |
       ...
```

If `start <= t < time`:

- Checkpoint 1 (`t_0 <= t < t_1`) includes:
  - measurement_0
- Checkpoint 2 (`t_1 <= t < t_2`) includes:
  - measurement_1, measurement_2, measurement_3

If `start < t <= time`:

- Checkpoint 1 (`t_0 < t <= t_1`) includes:
  - measurement_0, measurement_1
- Checkpoint 2 (`t_1 < t <= t_2`) includes:
  - measurement_2, measurement_3

(side note: with this interval definition, start is not included in any interval. This is not idea, but likely minor enough to not be a real problem in the real world)

@jmacd : From my understanding of you comment, you are saying that the `t_0 < t <= t_1` definition of the interval means that while Checkpoint 1 should contain both measurement_0 and measurement_1, it is not possible to do in practice. Mainly because the checkpoint process is no longer receiving measurements when `t_1` is defined by the process, right?

Reply:
https://github.com/OpenObservability/OpenMetrics/pull/126/files#diff-d1792782ae179ab49b581de250c5def2R103 

I talked to one of the StackDriver authors and they explained to me that they actually don't want to have inclusive interval because that may cause duplicate "report" (they also said that in practice nobody does implement that way - if an event happens at exactly same time as the endtime of the interval it will not be included in both deltas, the decision to be in the current delta or the next delta happens using other synchronization mechanisms that ensure a happens before relationship and ensures events are counted only once).

Reply:
> https://github.com/OpenObservability/OpenMetrics/pull/126/files#diff-d1792782ae179ab49b581de250c5def2R103
> 
> I talked to one of the StackDriver authors and they explained to me that they actually don't want to have inclusive interval because that may cause duplicate "report" (they also said that in practice nobody does implement that way - if an event happens at exactly same time as the endtime of the interval it will not be included in both deltas, the decision to be in the current delta or the next delta happens using other synchronization mechanisms that ensure a happens before relationship and ensures events are counted only once).

Thanks for the link and info. We should not consider a revision of an open PR and undocumented behavior given in a conversation to be prior art. Furthermore, I think the point @jmacd made is a technical one and I think we should engage the point on that level.

Reply:
@MrAlias I state back to my comment, using time to determine a happens-before relationship is impossible (or very hard to implement unless we are going to implement something like true-time). So I am not sure I understand your example, and how would it be possible for an event to happen before we start the checkpointer.

Reply:
@bogdandrutu: Hmm, it seems like we might be talking past each other. I'm wondering if this is stemming from the differences in concurrency design in the implementations we work in.

Maybe this is something we can talk about in the upcoming Metric SIG meeting? I'll add it to the agenda if you can make the meeting.